python/pyarrow/_cuda.pyx

```diff
@@ -246,6 +283,73 @@ cdef class Context:
             result.copy_from_device(buf, position=0, nbytes=size)
         return result
 
+    def buffer_from_object(self, obj):
+        """Create device buffer view of arbitray device memory object.
+
+        When the object contains a non-contiguous view of device
+        memory then the returned device buffer will contain contiguous
+        view of the memory, that is, including the intermediate data
+        that is otherwise invisible to the input object.
+
+        Parameters
+        ----------
+        obj : object
+          Specify object that contains address to device memory or
+          implements cuda array interface.
+
+        Returns
+        -------
+        cbuf : CudaBuffer
+          Device buffer as a view of device memory.
+
+        """
+        import numpy as np
+        if isinstance(obj, HostBuffer):
+            if not obj.address and obj.size == 0:
+                return self.new_buffer(0)
+            return self.foreign_buffer(obj.address, obj.size)
+        if isinstance(obj, Buffer):
+            obj = CudaBuffer.from_buffer(obj)
+        if isinstance(obj, CudaBuffer):
```

Comment:
`CudaBuffer` qualifies to being `buffer_from_object` argument and is useful for generic programming where the caller needs that its input can be transformed to `CudaBuffer` and applying `buffer_from_object` guarantees this. This is analogous to `numpy.asarray` that is applied to an object even if it is already an array.

Returning new copy of `CudaBuffer` might be theoretical but there is a possibility that the device address may be different in different contexts such as the current one and the context where the input `CudaBuffer` object was created. For instance, when one registers host memory, wraps it with `buffer_from_object`, then unregisters the host memory and subsequently registers the host memory again. I don't know if the device address of this memory will be the same as before. Hmm, if not then the former `CudaBuffer` object would contain invalid address because unregistering page-locked memory would invalidate the device address.

So, to make long story short, I'll presume that the above story represents a buggy design and returning `CudaBuffer` object unchanged is ok.

Reply:
Sounds good :-)

Reply:
Done.